### PR TITLE
feat(github): issue and pr lookup

### PR DESCRIPTION
**Pull request by almostSouji**
This PR adds the ability to look up issue, pr and commit data for arbitrary github repositories as well as allowing aliases to be set per guild and regex execution.

<details>
<summary>Tasks (initial release)</summary>

- [x] i18n: base translations
- [x] i18n: use native plural mechanism
- [x] i18n: make GH response data translatable
- [x] truncate embed fields based on embed limitations
- [x] use central utility for addField
- [x] validate repository name before making request to prevent invalid request body
- [x] github config command (separate)
- [x] simplify pr/issue output to just show relevant information
- [x] ditch flags, more information should be viewed in browser anyways
- [x] use configurable settings in pr command
- [x] i18n: rejections in issue-pr
- [x] config: truncate non-embeds before sending
- [x] regex: make execute on regex
- [x] regex: suppress errors when used as regex
- [x] rename command to `github` (alias `gh`)
- [x] incorporate alias setting into main cmd
- [x] issue lookup (incorporate into main cmd)
- [x] finalize non-configurable urls
- [x] remove section comment (indev util for easier file navigation)

</details>

### Future improvements

- [ ] 🚧 needs permission implementation | do not execute if missing embed permission)
- [ ] move query logic into API (Hasura action https://github.com/Naval-Base/yuudachi/pull/767#issuecomment-687714047)
- [ ] GH Oauth2 to support private repositories

### Usage

- `github alias <add|remove> <...alias:repo_identifier>` *
- `github <repository> <owner> <issue>` *
- `github <alias> <issue>`  *
- `<owner>/<repo>#<issue>` (non-prefixed) *
- `<alias>#<issue>` (non-prefixed) *

\* if issue is not a number try to resolve commit-like from it (different display and API call)   
\* repo identifier: `owner/repository` or a github link (SSH and HTTPS supported)

- 🚧 roadblock, requires other PR/changes/functionality

### CODE REVIEW

File: docker-compose.local.yml
Comment (iCrawl) [inline docker-compose.local.yml:82]: This shouldn't really be here should it?

File: src/handler/bin/index.ts
Comment (iCrawl) [inline src/handler/bin/index.ts:23]: A github token should not be a requirement to start the bot.

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:1]: I don't think the file should have `.types.` in its name, rather make a folder called `interfaces` or something.

Also: `GitHubAuthor`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:7]: `GitHubUser`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:11]: `GitHubLabel`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:17]: `GitHubCommit`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:21]: `GitHubReview`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:29]: `GitHubReviewState`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:37]: `GitHubAuthorAssociation`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:47]: `GitHubIssue`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:60]: `GitHubReviewDecision`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:66]: `GitHubPRData`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:79]: `GitHubPR`

File: src/handler/src/GitHub.types.ts
Comment (iCrawl) [inline src/handler/src/GitHub.types.ts:81]: `GitHubAPIData`

File: src/handler/src/commands/github/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/issue-pr.ts:58]: `ResultStatePR`

File: src/handler/src/commands/github/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/issue-pr.ts:118]: `PRIcons`

File: src/handler/src/commands/github/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/issue-pr.ts:156]: Should probably also use `Reflect` here instead of the `in` keyword.

**Review by iCrawl — CHANGES_REQUESTED** [review]

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:279]: This can be static.

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:371]: This can also be static.

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:400]: This can also be static.

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:413]: This can also be static.

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:417]: This can also be static.

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:421]: This can also be static.

File: src/handler/src/commands/github/issue-pr.ts
Comment (appellation) [inline src/handler/src/commands/github/issue-pr.ts:425]: This can also be static.

**Comment by appellation** [comment]
First of all, thank you for your contribution! I have some feedback on how I envision this being implemented; feel free to take as much or as little of it into consideration for this initial implementation, but I would like to see these changes for a final version.

I believe that this should be configurable on a per-guild basis. This means that we will have to hook into Postgres and pull in repository configuration for the guild. I think we'll only have to store a list of configured repository URLs for each guild, so this should be a pretty straightforward change to the schema.

I also think that all of this complex GitHub query logic should be moved into the API; there is a strong likelihood of needing this data elsewhere in the application, and we really want to reduce code duplication. This would also entail adding a [remote schema](https://hasura.io/docs/1.0/graphql/core/remote-schemas/index.html) to Hasura for use in the frontend.

We should also consider supporting private repositories, which means storing GitHub OAuth2 credentials. Since the frontend is pretty non-existent at this point, it could be kind of hard to test: this might be something we want to put off until we can actually link a GitHub account to a guild/user. However, by setting this up with per-guild configuration *and* moving the query logic to the API, it should be easy to use existing API middleware and authorization logic to validate access and protect user data.

Thanks again for starting work on this feature! Let me know if I can better clarify any of my comments.

**Comment by almostSouji** [comment]
### Configuration:

Also requires the labels to be configurable (one emoji per color, colors and emoji differ across guilds, although a grey default does exist).

Badge can be made configurable as well, thought doesn't necessarily need to if moved to a more general motif then the discord.js logo (since this was originally implemented as djs only feature)

Per guild setup (probably best in separate command, maybe consider add/remove/set for better UX):
- allowed repositories
- label emojis

### Oauth:

This might make it viable to include a check (if the permissions are present) for the badge only to be applied to maintainers which actually count towards the review requirements (which was the initial thought process behind that feature, which I moved towards author assoc. which at most can discern members of the org)

File: src/handler/src/commands/github/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/issue-pr.ts:146]: You can still throw here, look at the tag command for that.

File: src/handler/src/commands/github/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/issue-pr.ts:443]: There is already functionality for plurals in i18next:
https://www.i18next.com/translation-function/plurals

Maybe this can be useful.

**Comment by appellation** [comment]
Regarding permissions checks, you can consider using the Spectacles util class that enables simpler permission checking.

https://github.com/spec-tacles/spectacles.js/tree/master/packages/util#permissions

File: src/handler/package.json
Comment (NotSugden) [inline src/handler/package.json:22]: this could be updated to 2.6.1

File: src/handler/src/commands/github/issue-pr.ts
Comment (NotSugden) [inline src/handler/src/commands/github/issue-pr.ts:270]:
```suggestion
					const reviewState = IssuePRLookup.cleanDecision(r.state);
```
function returns a string

File: src/handler/src/commands/github/issue-pr.ts
Comment (NotSugden) [inline src/handler/src/commands/github/issue-pr.ts:303]: escape quote marks/newlines in the `respository`? to prevent a malformed request if a user put in quotemarks

Comment (almostSouji) [reply src/handler/src/commands/github/issue-pr.ts:315]: opted for validating the name instead of escaping in https://github.com/Naval-Base/yuudachi/pull/767/commits/147f70847a93da6643b19bebd9590ef5e0ed58ab . regex taken from https://stackoverflow.com/a/59082561, approach seems valid enough from a bit of own testing and reading linked moby issue

File: src/handler/src/commands/github/issue-pr.ts
Comment (NotSugden) [inline src/handler/src/commands/github/issue-pr.ts:129]:
```suggestion
	OPEN = 'https://cdn.discordapp.com/emojis/751210140086042686.png',
```
consistency with other urls

Comment (almostSouji) [reply src/handler/src/commands/github/issue-pr.ts:129]: urls and emojis will change before merge anyways

File: src/api/public/locales/en/handler.json
Comment (iCrawl) [inline src/api/public/locales/en/handler.json:127]: This file should have a newline at the end.

File: src/handler/src/commands/github/sub/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/issue-pr.ts:62]: `const res: GitHubAPIResult = ...` instead of `as GitHubAPIResult`

File: src/handler/src/commands/github/sub/issue-pr.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/issue-pr.ts:208]: You probably want to await this?

File: src/handler/src/commands/github/sub/commit.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/commit.ts:69]: You probably want to await this?

File: src/handler/src/commands/github/sub/commit.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/commit.ts:23]: `const res: GitHubAPIResult = ...` instead of `as GitHubAPIResult`

File: src/handler/src/commands/github/sub/alias.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/alias.ts:141]: Un-indent those 1 time.

File: src/handler/src/commands/github/sub/alias.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/alias.ts:180]: Un-indent those 1 time.

File: src/handler/src/commands/github/sub/alias.ts
Comment (iCrawl) [inline src/handler/src/commands/github/sub/alias.ts:90]: Un-indent this 1 time.

**Review by iCrawl — APPROVED** [review]

**Review by appellation — APPROVED** [review]
I'm not a huge fan of having functions with a lot of arguments, particularly when none of them are documented. However, I do like the approach you've taken here with splitting things up a bit more. I'm wondering if any of this logic could be abstracted out of the commands section and into general source code, but this seems like a reasonable base from which to build on.

**Pull request merged**
